python/ray/tests/kuberay/test_autoscaling_e2e.py

```diff
@@ -138,6 +122,30 @@ def _get_ray_cr_config(
         gpu_group["groupName"] = "fake-gpu-group"
         config["spec"]["workerGroupSpecs"].append(gpu_group)
 
+        # Substitute images.
+        for group_spec in config["spec"]["workerGroupSpecs"] + [
+            config["spec"]["headGroupSpec"]
+        ]:
+            containers = group_spec["template"]["spec"]["containers"]
+
+            # Will assume the example config maintains this invariant:
+            ray_container = containers[0]
+
+            ray_container["image"] = RAY_IMAGE
+
+            for container in containers:
+                container["imagePullPolicy"] = PULL_POLICY
+
+        head_containers = config["spec"]["headGroupSpec"]["template"]["spec"][
+            "containers"
+        ]
+        autoscaler_container = [
```

Comment:
Very interesting idiom!

Reply:
a for loop would have worked just as well... main function of a math phd is to induce weird programming habits